### PR TITLE
populate PID for container entity on containerd runtime

### PR DESCRIPTION
**Pull request by adel121**
### What does this PR do?

Populates the PID field of container entity of workloadmeta in containerd runtime. 

### Motivation

Currently, the containerd collector sets PID field to 0 by default for all container entities that it generates.
We should be able to populate this field.

### Additional Notes

- For containers that are not in the running state, the PID field will be 0.

### Possible Drawbacks / Trade-offs

None

### Describe how to test/QA your changes

Deploy the agent in containerd runtime environment (example, on kind cluster), and verify that container entities of running containers get the PID field correctly populated.

Ensure that PID is 0 for non running containers, and non zero for running containers.

Example:

```
agent workload-list

=== Entity container sources(merged):[node_orchestrator runtime] id: f4a6cadb8c218e6bcc665de99a50050f31018aa4213d4daec27017ba355004b2 ===
----------- Entity ID -----------
Kind: container ID: f4a6cadb8c218e6bcc665de99a50050f31018aa4213d4daec27017ba355004b2
----------- Entity Meta -----------
Name: coredns
Namespace: k8s.io
Annotations: 
Labels: io.cri-containerd.kind:container io.kubernetes.container.name:coredns io.kubernetes.pod.name:coredns-5d78c9869d-zqwzh io.kubernetes.pod.uid:ba28258a-b68d-4f3b-ae07-dbc677af1dee io.kubernetes.pod.namespace:kube-system 
----------- Image -----------
Name: registry.k8s.io/coredns/coredns
Tag: v1.10.1
ID: sha256:97e04611ad43405a2e5863ae17c6f1bc9181bdefdaa78627c432ef754a4eb108
Raw Name: registry.k8s.io/coredns/coredns:v1.10.1
Short Name: coredns
Repo Digest: sha256:a0ead06651cf580044aeb0a0feba63591858fb2e43ade8c9dea45a6a89ae7e5e
----------- Container Info -----------
Runtime: containerd
RuntimeFlavor: 
Running: true
Status: running
Health: 
Created At: 2024-06-03 08:55:32.257184715 +0000 UTC
Started At: 2024-06-03 08:55:32.257184715 +0000 UTC
Finished At: 0001-01-01 00:00:00 +0000 UTC
----------- Resources -----------
TargetCPUUsage: 10
TargetMemoryUsage: 73400320
Allowed env variables: 
Hostname: 
Network IPs: :10.244.0.3 
PID: 1628
Cgroup path: kubelet-kubepods-burstable-podba28258a_b68d_4f3b_ae07_dbc677af1dee.slice/cri-containerd-f4a6cadb8c218e6bcc665de99a50050f31018aa4213d4daec27017ba355004b2.scope
----------- Ports -----------
Port: 53
Name: dns
Protocol: UDP
Host Port: 0
Port: 53
Name: dns-tcp
Protocol: TCP
Host Port: 0
Port: 9153
Name: metrics
Protocol: TCP
Host Port: 0
----------- Security Context -----------
----------- Capabilities -----------
Add: [NET_BIND_SERVICE]
Drop: [all]
Privileged: false
===

=== Entity container sources(merged):[runtime] id: 28ef5afd89655116febe6fea719cace2c2d8f14a8f1897445c10be14852221b9 ===
----------- Entity ID -----------
Kind: container ID: 28ef5afd89655116febe6fea719cace2c2d8f14a8f1897445c10be14852221b9
----------- Entity Meta -----------
Name: 
Namespace: k8s.io
Annotations: 
Labels: io.kubernetes.pod.uid:***************************e78ae maintainers:Kubernetes Authors description:go based runner for distroless scenarios io.cri-containerd.kind:container io.kubernetes.container.name:kube-apiserver io.kubernetes.pod.name:kube-apiserver-kind-control-plane io.kubernetes.pod.namespace:kube-system 
----------- Image -----------
Name: registry.k8s.io/kube-apiserver
Tag: v1.27.3
ID: sha256:634c53edb5c14785aa9360617dcf173dc8f56081eff42a0d0da8ad823cc7773f
Raw Name: registry.k8s.io/kube-apiserver:v1.27.3
Short Name: kube-apiserver
Repo Digest: sha256:024134bb4e61a4975c5c15db76a7e3571eea74eff90c12a730781bf9c9daedd0
----------- Container Info -----------
Runtime: containerd
RuntimeFlavor: 
Running: false
Status: unknown
Health: 
Created At: 2024-05-30 13:29:06.768518043 +0000 UTC
Started At: 2024-05-30 13:29:06.768518043 +0000 UTC
Finished At: 0001-01-01 00:00:00 +0000 UTC
----------- Resources -----------
Allowed env variables: 
Hostname: 
Network IPs: 
PID: 0
Cgroup path: kubelet-kubepods-burstable-podb97232df7635c1e88c7bd9309b8e78ae.slice/cri-containerd-28ef5afd89655116febe6fea719cace2c2d8f14a8f1897445c10be14852221b9.scope
===
```

We can see that PID is set to 1628 for the running container, and 0 for the second container which has a status `UNKNOWN`

### CODE REVIEW

File: comp/core/workloadmeta/collectors/internal/containerd/container_builder.go [hunk 1]
```diff
@@ -57,6 +57,15 @@ func buildWorkloadMetaContainer(namespace string, container containerd.Container
 		}
 	}
 
+	// Get Container PID
+	pid := 0
+	task, err := container.Task(ctx, nil)
```
[the remark this answers is not in the document]
Comment (adel121) [reply comp/core/workloadmeta/collectors/internal/containerd/container_builder.go:62]: I used [the same context](https://github.com/DataDog/datadog-agent/blob/4335d8244ffefb11e693c2669221de18d303fbd2/comp/core/workloadmeta/collectors/internal/containerd/container_builder.go#L45-L46) used for other calls in this collector. 

By checking usages of this API in containerd repo, I see that using a context with timeout is doable. Here is an example:

1. Client, Context and Cancel channel are initialised [here](https://github.com/containerd/containerd/blob/1369112514d7200d9945cad883873a6f0ce433a4/cmd/ctr/commands/containers/checkpoint.go#L56) (The context is created [here](https://github.com/containerd/containerd/blob/1369112514d7200d9945cad883873a6f0ce433a4/cmd/ctr/commands/client.go#L60)).
2. The timed context is used [here](https://github.com/containerd/containerd/blob/1369112514d7200d9945cad883873a6f0ce433a4/cmd/ctr/commands/containers/checkpoint.go#L78).


Do you think we should refactor this to use a context with timeout?

File: comp/core/workloadmeta/collectors/internal/containerd/container_builder.go [hunk 2]
```diff
@@ -57,6 +57,16 @@ func buildWorkloadMetaContainer(namespace string, container containerd.Container
 		}
 	}
 
+	// Get Container PID
+	var pid int
+	task, err := container.Task(ctx, nil)
+	if err == nil {
+		pid = int(task.Pid())
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/containerd/container_builder.go:64]: what is the origin type of the Pid() ? uint?

**Review by clamoriniere — APPROVED** [review]

File: comp/core/workloadmeta/collectors/internal/containerd/container_builder.go (hunk 2 above)
Comment (adel121) [reply comp/core/workloadmeta/collectors/internal/containerd/container_builder.go:64]: yes

File: comp/core/workloadmeta/collectors/internal/containerd/container_builder_test.go [hunk 3]
```diff
@@ -9,6 +9,8 @@ package containerd
 
 import (
 	"context"
+	"fmt"
+	"github.com/containerd/containerd/cio"
```
Comment (clamoriniere) [inline comp/core/workloadmeta/collectors/internal/containerd/container_builder_test.go:13]: could you move the `github.com/containerd/containerd/cio` outside the golang official package

**Review by clamoriniere — APPROVED** [review]

**Comment by songy23** [comment]
Hi @adel121 I was assigned a QA card for this PR and wonder why is this marked as `team/opentelemetry`? Doesn't seem this PR has to do with OTel.

Could you move https://datadoghq.atlassian.net/browse/OTEL-1874 to the correct team if ^ is the case?

**Comment by adel121** [comment]
Hi @songy23 

Yes indeed it should not be for team/opentelemetry.

Looks like it was added automatically by the github bot due to a commit that touched a file owned by otel and which was reverted after that, but the team/opentelemetry wan't removed then. 

I will move it to the right team, thanks.

**Pull request merged**